hnn_core/lfp.py

```diff
@@ -0,0 +1,243 @@
+        b = ele_pos - end
+        # projection: H = a.cos(theta) = a.dot(b) / |a|
+        H = np.dot(b, a) / norm_a  # NB can be negative
+        L = H + norm_a
+        R2 = np.dot(b, b) - H ** 2  # NB squares
+        # To avoid numerical errors when electrode is placed (anywhere) on the
+        # section axis, enforce minimal axial distance
+        R2 = max(R2, (section.diam / 2.0 + 0.1) ** 2)
+
+        if L < 0 and H < 0:  # electrode is "behind" section
+            num = np.sqrt(H ** 2 + R2) - H  # == norm(b) - H
+            denom = np.sqrt(L ** 2 + R2) - L
+        elif L > 0 and H < 0:  # electrode is "on top of" section
+            num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+            denom = R2
+        else:  # electrode is "ahead of" section
+            num = np.sqrt(L ** 2 + R2) + L
+            denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+        phi = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class _LFPElectrode:
+    """Local field potential (LFP) electrode class.
+
+    Parameters
+    ----------
+    coord : tuple
+        The (x, y, z) coordinates (in um) of the LFP electrode.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    pc : instance of h.ParallelContext()
+        ParallelContext instance for running in parallel
+    cvode : instanse of h.CVode
+        Multi order variable time step integration method.
+    method : str
+        'psa' (default), i.e., point source approximation or line source
+        approximation, i.e., 'lsa'
+
+    Attributes
+    ----------
+    lfp_t : instance of h.Vector
+        The time points (in ms) at which the LFP is calculated.
+    lfp_v : instance of h.Vector
+        The LFP voltage (in uV).
+
+    Notes
+    -----
+    See Table 5 in http://jn.physiology.org/content/104/6/3388.long for
+    measured values of sigma in rat cortex (note units there are mS/cm)
+    """
+
+    def __init__(self, coord, sigma=0.3, pc=None, cvode=None, method='psa'):
+
+        # ordered list of h.Sections on this rank (if running in parallel)
+        secs_on_rank = _get_sections_on_this_rank()
+        # np.array of number of segments for each section, ordered as above
+        segment_counts = np.array(_get_segment_counts(secs_on_rank))
+
+        # pointers assigned to _ref_i_membrane_ at each EACH internal segment
+        self.imem_ptrvec = h.PtrVector(segment_counts.sum())
+        # placeholder into which pointer values are read on each sim step
+        self.imem_vec = h.Vector(int(self.imem_ptrvec.size()))
```

Comment:
I think this may be an area for a speed improvement. Since a different instance of `_LFP_Electrode` is instantiated for every electrode, an identical `imem_vec` is present for every instance. An alternative would be to have the class accept a list of electrode positions, and loop through the transfer resistances for the dot product in `_LFP_Electrode.callback`.

Reply:
Way ahead of you ;) You're gonna love my next commits, I hope!